cpp/arcticdb/storage/test/test_storage_exceptions.cpp

```diff
@@ -0,0 +1,164 @@
+        cfg.set_map_size(map_size);
+        cfg.set_recreate_if_exists(true);
+        cfg.set_recreate_if_exists(true);
+
+        acs::LibraryPath library_path{"a", "b"};
+
+        return std::make_unique<acs::lmdb::LmdbStorage>(library_path, acs::OpenMode::WRITE, cfg);
+    }
+};
+
+// Generic tests that run with all types of storages
+
+class GenericStorageTest : public ::testing::TestWithParam<std::shared_ptr<StorageFactory>> {
+protected:
+    std::unique_ptr<acs::Storage> storage;
+
+    void SetUp() override {
+        if (!fs::exists(TEST_DATABASES_PATH)) {
+            fs::create_directories(TEST_DATABASES_PATH);
+        }
+        storage = GetParam()->Create();
+    }
+
+    void TearDown() override {
+        if (fs::exists(TEST_DATABASES_PATH)) {
+            fs::remove_all(TEST_DATABASES_PATH);
+        }
+
+        storage = nullptr;
+    }
+};
+
+TEST_P(GenericStorageTest, WriteDuplicateKeyException) {
+    ac::entity::AtomKey k = ac::entity::atom_key_builder().gen_id(0).build<ac::entity::KeyType::VERSION>("sym");
+
+    acs::KeySegmentPair kv(k);
+    kv.segment().set_buffer(std::make_shared<ac::Buffer>());
+
+    storage->write(std::move(kv));
+
+    ASSERT_TRUE(storage->key_exists(k));
+
+    acs::KeySegmentPair kv1(k);
+    kv1.segment().set_buffer(std::make_shared<ac::Buffer>());
+
+
+    ASSERT_THROW({
+        storage->write(std::move(kv1));
+    },  acs::DuplicateKeyException);
+
+}
+
+TEST_P(GenericStorageTest, ReadKeyNotFoundException) {
+    ac::entity::AtomKey k = ac::entity::atom_key_builder().gen_id(0).build<ac::entity::KeyType::VERSION>("sym");
+
+    ASSERT_TRUE(!storage->key_exists(k));
+    ASSERT_THROW({
+                    storage->read(k, acs::ReadKeyOpts{});
+                 }, acs::KeyNotFoundException);
+
+}
+
+TEST_P(GenericStorageTest, UpdateKeyNotFoundException) {
+    ac::entity::AtomKey k = ac::entity::atom_key_builder().gen_id(0).build<ac::entity::KeyType::VERSION>("sym");
+
+    acs::KeySegmentPair kv(k);
+    kv.segment().header().set_start_ts(1234);
+    kv.segment().set_buffer(std::make_shared<ac::Buffer>());
+
+    ASSERT_TRUE(!storage->key_exists(k));
+    ASSERT_THROW({
+                    storage->update(std::move(kv), acs::UpdateOpts{});
+                 }, acs::KeyNotFoundException);
+
+}
+
+TEST_P(GenericStorageTest, RemoveKeyNotFoundException) {
+    ac::entity::AtomKey k = ac::entity::atom_key_builder().gen_id(0).build<ac::entity::KeyType::VERSION>("sym");
+
+    ASSERT_TRUE(!storage->key_exists(k));
+    ASSERT_THROW({
+                     storage->remove(k, acs::RemoveOpts{});
+                 }, acs::KeyNotFoundException);
+
+}
+
+INSTANTIATE_TEST_SUITE_P(
+        AllStoragesCommonTests,
+        GenericStorageTest,
+        ::testing::Values(std::make_shared<LMDBStorageFactory>())
+);
+
+// LMDB Storage specific tests
+
+class LMDBStorageTestBase : public ::testing::TestWithParam<std::shared_ptr<LMDBStorageFactory>> {
```

Comment:
This is used in `TEST_F` suites, not `TEST_P`. It should be possible to remove the `::testing::TestWithParam` then.